### ISSUE TITLE
foomatic-rip segfaults with 'job-sheets=none,none' but works with 'job-sheets=none'

### ISSUE THREAD

**Issue by jsmeix** [OpenPrinting/cups-filters#348]
I have cups-filters version 1.27.2
but - as far as I see - the issue is likely also in current master code.

How to reproduce:
```
# export PPD=/etc/cups/ppd/ps.ppd
# export CUPS_DATADIR=/usr/share/cups
# export CUPS_CACHEDIR=/var/cache/cups

# cat /usr/share/ghostscript/9.52/examples/colorcir.ps \
 | /usr/lib/cups/filter/foomatic-rip job user title 1 'job-sheets=none,none'

DEBUG: Color Manager: Invalid printer name.
'CM Color Calibration' Mode in SPOOLER-LESS: Off
Getting input from file 
foomatic-rip version 1.27.2 running...
Parsing PPD file ...
Added option ColorSpace
Added option PageSize
Added option ImageableArea
Added option PaperDimension
Added option InputSlot
Added option Resolution
Added option Duplex
Added option Font

Parameter Summary
-----------------

Spooler: cups
Printer: unknown
Shell: /bin/bash
Printer model: Generic postscript printer
Job title: title
File(s) to be printed:
<STDIN>

Ghostscript extra search path ('GS_LIB'): /usr/share/cups/fonts
Printing system options:
Pondering option 'job-sheets=none'
Unknown option job-sheets=none.
Pondering option 'none'
CM Color Calibration Mode in CUPS: Off
Options from the PPD file:

================================================

File: <STDIN>

================================================

Filetype: PostScript
Storing temporary files in /tmp
File contains 1 pages.
Reading PostScript input ...
"Prolog" section is missing, inserting it.
"Setup" section is missing, inserting it.
"PageSetup" section is missing, inserting it.
Inserting option code into "PageSetup" section.
Segmentation fault (core dumped)
```

Versus (leaving out most of the PostScript at `[...]`)
```
# cat /usr/share/ghostscript/9.52/examples/colorcir.ps \
 | /usr/lib/cups/filter/foomatic-rip job user title 1 'job-sheets=none'

DEBUG: Color Manager: Invalid printer name.
'CM Color Calibration' Mode in SPOOLER-LESS: Off
Getting input from file 
foomatic-rip version 1.27.2 running...
Parsing PPD file ...
Added option ColorSpace
Added option PageSize
Added option ImageableArea
Added option PaperDimension
Added option InputSlot
Added option Resolution
Added option Duplex
Added option Font

Parameter Summary
-----------------

Spooler: cups
Printer: unknown
Shell: /bin/bash
PPD file: /etc/cups/ppd/ps.ppd
ATTR file: 
Printer model: Generic postscript printer
Job title: title
File(s) to be printed:
<STDIN>

Ghostscript extra search path ('GS_LIB'): /usr/share/cups/fonts
Printing system options:
Pondering option 'job-sheets=none'
Unknown option job-sheets=none.
CM Color Calibration Mode in CUPS: Off
Options from the PPD file:

================================================

File: <STDIN>

================================================

Filetype: PostScript
Storing temporary files in /tmp
File contains 1 pages.
Reading PostScript input ...
"Prolog" section is missing, inserting it.
"Setup" section is missing, inserting it.
"PageSetup" section is missing, inserting it.
Inserting option code into "PageSetup" section.
Flushing FIFO.

Starting renderer with command: "cat"
Starting process "kid3" (generation 1)

Closing renderer
Starting process "kid4" (generation 2)
Starting process "renderer" (generation 2)
JCL: 2345X@PJL
<job data> 

renderer exited with status 0
%!
%%BeginProlog
%%EndProlog
%%BeginSetup
%%EndSetup
%%BeginPageSetup
%%EndPageSetup
/colorcirsave save def          % prevent left over effects

[...]

graycircle

showpage
clear cleardictstack
colorcirsave restore
kid4 exited with status 0
kid3 finished
kid3 exited with status 0

Closing foomatic-rip.

# echo $?
0
```

As far as we (i.e. a colleague at SUSE) found out the reason is the following:

At
https://github.com/OpenPrinting/cups-filters/blob/master/filter/foomatic-rip/options.c#L851
```
    char **paramvalues = paramvalues_from_string(opt, values);

    dstrclear(cmd);
    for (param = opt->paramlist, i = 0; param; param = param->next, i++)
        dstrcatf(cmd, "%s ", paramvalues[i]);
```
the call to paramvalues_from_string is returning null
and accessing paramvalues[i] causes a crash.

Our (i.e. the colleague at SUSE) minimal ad-hoc in-place fix is therefore
```
     char **paramvalues = paramvalues_from_string(opt, values);
 
     dstrclear(cmd);
-    for (param = opt->paramlist, i = 0; param; param = param->next, i++)
-        dstrcatf(cmd, "%s ", paramvalues[i]);
+    if (paramvalues)
+    {
+        for (param = opt->paramlist, i = 0; param; param = param->next, i++)
+            if (paramvalues[i])
+                dstrcatf(cmd, "%s ", paramvalues[i]);
+    }
     dstrcat(cmd, opt->custom_command);
     free_paramvalues(opt, paramvalues);
 }
```
to only avoid the crash (which is in general good to do).

A real fix would be to properly recognize the
`job-sheets=none,none` in the option parser.

**Comment by jsmeix** [comment]
Same kind of code is also at
https://github.com/OpenPrinting/cups-filters/blob/master/filter/foomatic-rip/options.c#L860
```
void build_cups_custom_jcl_command(dstr_t *cmd, option_t *opt, const char *values)
...
    char **paramvalues = paramvalues_from_string(opt, values);
```
so it looks like a more general problem
perhaps in `paramvalues_from_string`?

**Comment by tillkamppeter** [comment]
Fixed in commit 14a15d5 (master) and commit 67c2128 (1.x).
Thank you for the bug report.
